File: pdf_export.py
import sys
import os
import re
import json
import fitz  # PyMuPDF
import platform
from html.parser import HTMLParser


# ---------------------------------------------------------------------------
# HTML → Plain Text Converter
# ---------------------------------------------------------------------------

class _HtmlSegment:
    """Represents a run of text with an optional style hint."""
    __slots__ = ('text', 'bold', 'italic')

    def __init__(self, text: str, bold: bool = False, italic: bool = False):
        self.text   = text
        self.bold   = bold
        self.italic = italic


class HtmlToTextConverter(HTMLParser):
    """
    Converts a subset of HTML to plain text suitable for PDF insertion.

    Supported tags
    --------------
    Block-level  : <p>, <div>, <br>, <li>, <tr>, <h1>–<h6>
    Inline-style  : <b>, <strong>, <i>, <em>   (tracked but rendered as plain text)
    Ignored       : <style>, <script>, <head> and all unknown tags

    The converter collapses consecutive blank lines to a single blank line
    and trims leading/trailing whitespace from the final result.
    """

    _BLOCK_TAGS   = frozenset({'p', 'div', 'li', 'tr', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'})
    _SKIP_TAGS    = frozenset({'style', 'script', 'head'})
    _BOLD_TAGS    = frozenset({'b', 'strong'})
    _ITALIC_TAGS  = frozenset({'i', 'em'})

    def __init__(self):
        super().__init__()
        self._parts: list[str] = []
        self._skip_depth  = 0
        self._bold_depth  = 0
        self._italic_depth = 0

    # ------------------------------------------------------------------ #
    # HTMLParser callbacks                                                 #
    # ------------------------------------------------------------------ #

    def handle_starttag(self, tag: str, attrs):
        tag = tag.lower()
        if tag in self._SKIP_TAGS:
            self._skip_depth += 1
            return
        if self._skip_depth:
            return
        if tag in self._BLOCK_TAGS:
            self._parts.append('\n')
        elif tag == 'br':
            self._parts.append('\n')
        if tag in self._BOLD_TAGS:
            self._bold_depth += 1
        if tag in self._ITALIC_TAGS:
            self._italic_depth += 1

    def handle_endtag(self, tag: str):
        tag = tag.lower()
        if tag in self._SKIP_TAGS:
            self._skip_depth = max(0, self._skip_depth - 1)
            return
        if self._skip_depth:
            return
        if tag in self._BLOCK_TAGS:
            self._parts.append('\n')
        if tag in self._BOLD_TAGS:
            self._bold_depth = max(0, self._bold_depth - 1)
        if tag in self._ITALIC_TAGS:
            self._italic_depth = max(0, self._italic_depth - 1)

    def handle_data(self, data: str):
        if not self._skip_depth:
            self._parts.append(data)

    def handle_entityref(self, name: str):
        entities = {'amp': '&', 'lt': '<', 'gt': '>', 'nbsp': ' ',
                    'quot': '"', 'apos': "'"}
        if not self._skip_depth:
            self._parts.append(entities.get(name, ''))

    def handle_charref(self, name: str):
        if self._skip_depth:
            return
        try:
            code = int(name[1:], 16) if name.startswith(('x', 'X')) else int(name)
            self._parts.append(chr(code))
        except (ValueError, OverflowError):
            pass

    # ------------------------------------------------------------------ #
    # Public API                                                           #
    # ------------------------------------------------------------------ #

    def get_text(self) -> str:
        text = ''.join(self._parts)
        # Collapse 3+ consecutive newlines → 2
        text = re.sub(r'\n{3,}', '\n\n', text)
        return text.strip()


def html_to_plain_text(raw: str) -> str:
    """
    Convert an HTML string to plain text.  If the string contains no HTML
    tags at all it is returned unchanged (fast path).
    """
    if '<' not in raw:
        return raw
    converter = HtmlToTextConverter()
    converter.feed(raw)
    return converter.get_text()


# ---------------------------------------------------------------------------
# Font discovery
# ---------------------------------------------------------------------------

def get_system_font_paths() -> list[str]:
    """
    Return a priority-ordered list of font file paths that support
    Unicode / Cyrillic characters.  The first existing path is used.
    """
    system = platform.system()

    if system == 'Windows':
        return [
            r'C:\Windows\Fonts\arial.ttf',
            r'C:\Windows\Fonts\arialuni.ttf',   # Arial Unicode MS
            r'C:\Windows\Fonts\calibri.ttf',
            r'C:\Windows\Fonts\tahoma.ttf',
        ]
    if system == 'Darwin':
        return [
            '/Library/Fonts/Arial.ttf',
            '/System/Library/Fonts/Supplemental/Arial.ttf',
            '/Library/Fonts/Microsoft/Arial.ttf',
            '/System/Library/Fonts/Helvetica.ttc',
        ]
    # Linux / other
    return [
        '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',
        '/usr/share/fonts/dejavu-sans-fonts/DejaVuSans.ttf',
        '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf',
        '/usr/share/fonts/liberation/LiberationSans-Regular.ttf',
        '/usr/share/fonts/TTF/arial.ttf',
        '/usr/share/fonts/truetype/freefont/FreeSans.ttf',
    ]


def find_system_font() -> str | None:
    """Return the first usable Unicode font path, or None."""
    for path in get_system_font_paths():
        if os.path.isfile(path):
            return path
    return None


# ---------------------------------------------------------------------------
# Color helper
# ---------------------------------------------------------------------------

def parse_color(hex_str: str) -> tuple[float, float, float]:
    """Convert ``#RGB`` or ``#RRGGBB`` to a (r, g, b) tuple in [0, 1]."""
    try:
        s = hex_str.lstrip('#')
        if len(s) == 3:
            s = ''.join(c * 2 for c in s)
        return (
            int(s[0:2], 16) / 255.0,
            int(s[2:4], 16) / 255.0,
            int(s[4:6], 16) / 255.0,
        )
    except Exception:
        return (0.0, 0.0, 0.0)


# ---------------------------------------------------------------------------
# Font-size auto-fit
# ---------------------------------------------------------------------------

_FIT_SCALE_STEPS = (1.0, 0.9, 0.8, 0.7, 0.6, 0.5)

def insert_text_fitted(page: fitz.Page, rect: fitz.Rect, text: str,
                       fontsize: float, color, font_kwargs: dict) -> bool:
    """
    Try to insert *text* into *rect* at *fontsize*.  If it doesn't fit,
    progressively shrink the font size through ``_FIT_SCALE_STEPS`` until
    it fits or all steps are exhausted.

    Returns True if the text was successfully inserted at some size.
    """
    for scale in _FIT_SCALE_STEPS:
        fs = fontsize * scale
        kwargs = {**font_kwargs, 'fontsize': fs, 'color': color, 'align': 0}
        result = page.insert_textbox(rect, text, **kwargs)
        if result >= 0:
            return True
    # Last-ditch attempt at minimum size — insert anyway (text may clip)
    kwargs = {**font_kwargs, 'fontsize': fontsize * _FIT_SCALE_STEPS[-1],
              'color': color, 'align': 0}
    page.insert_textbox(rect, text, **kwargs)
    return False


# ---------------------------------------------------------------------------
# Core processing
# ---------------------------------------------------------------------------

def process_pdf(source_path: str, output_path: str, data: dict) -> None:
    # ── Open source document ────────────────────────────────────────────
    try:
        doc = fitz.open(source_path)
    except Exception as exc:
        sys.stderr.write(f'[pdf_export] Error opening source PDF: {exc}\n')
        sys.exit(1)

    # ── Locate a Unicode-capable font ──────────────────────────────────
    font_path = find_system_font()
    font_key  = 'uf'          # arbitrary internal name for PyMuPDF

    if font_path:
        print(f'[pdf_export] Unicode font: {font_path}')
    else:
        print('[pdf_export] WARNING: No Unicode font found — '
              'Cyrillic/non-Latin characters may appear as "????". '
              'Install DejaVu Sans or Arial to fix this.')

    # Build font kwargs once so we don't repeat the dict every iteration
    base_font_kwargs: dict = {}
    if font_path:
        base_font_kwargs['fontname'] = font_key
        base_font_kwargs['fontfile'] = font_path
    else:
        base_font_kwargs['fontname'] = 'helv'

    total_overlays = sum(len(v) for v in data.values())
    print(f'[pdf_export] Processing {len(data)} page(s), '
          f'{total_overlays} overlay(s)…')

    # ── Page loop ───────────────────────────────────────────────────────
    for page_num_str, overlays in data.items():
        try:
            page_idx = int(page_num_str) - 1          # 1-based → 0-based
        except ValueError:
            print(f'[pdf_export] Skipping invalid page key: {page_num_str!r}')
            continue

        if page_idx < 0 or page_idx >= doc.page_count:
            print(f'[pdf_export] Page {page_num_str} out of range '
                  f'(doc has {doc.page_count} pages) — skipped.')
            continue

        page   = doc[page_idx]
        page_w = page.rect.width
        page_h = page.rect.height

        for item in overlays:
            # ── Resolve and sanitize text ────────────────────────────
            raw_text = item.get('text', '')
            text = html_to_plain_text(raw_text).strip()
            if not text:
                continue

            # ── Coordinates ─────────────────────────────────────────
            try:
                x = float(item.get('x', 0))
                y = float(item.get('y', 0))
                w = float(item.get('width',  0))
                h = float(item.get('height', 0))
            except (TypeError, ValueError):
                print(f'[pdf_export] Bad coordinates on page {page_num_str} — skipped.')
                continue

            if item.get('isNormalized', False):
                rect = fitz.Rect(
                    x       * page_w,
                    y       * page_h,
                    (x + w) * page_w,
                    (y + h) * page_h,
                )
            else:
                rect = fitz.Rect(x, y, x + w, y + h)

            # Guard against degenerate rectangles
            if rect.is_empty or rect.width <= 1 or rect.height <= 1:
                print(f'[pdf_export] Degenerate rect on page {page_num_str} '
                      f'({rect}) — skipped.')
                continue

            # ── Font size ────────────────────────────────────────────
            fs_raw = item.get('fontSize', [10])
            if not isinstance(fs_raw, (list, tuple)):
                fs_raw = [fs_raw]
            try:
                fontsize = max(4.0, float(fs_raw[0]))
            except (TypeError, ValueError, IndexError):
                fontsize = 10.0

            # ── Color ────────────────────────────────────────────────
            color = parse_color(item.get('color', '#000000'))

            # ── Draw white background (erase original content) ───────
            bg = fitz.Rect(rect.x0 - 1, rect.y0 - 1,
                           rect.x1 + 1, rect.y1 + 1)
            page.draw_rect(bg, color=(1, 1, 1), fill=(1, 1, 1), overlay=True)

            # ── Insert text with auto-fit ────────────────────────────
            try:
                fitted = insert_text_fitted(
                    page, rect, text,
                    fontsize, color, base_font_kwargs,
                )
                if not fitted:
                    print(f'[pdf_export] Text clipped on page {page_num_str}: '
                          f'{text[:40]!r}…')
            except Exception as exc:
                print(f'[pdf_export] Warning — could not draw text on '
                      f'page {page_num_str}: {exc}')

    # ── Save ────────────────────────────────────────────────────────────
    try:
        doc.save(output_path, garbage=4, deflate=True)
        doc.close()
        print(f'[pdf_export] Saved → {output_path}')
    except Exception as exc:
        sys.stderr.write(f'[pdf_export] Error saving PDF: {exc}\n')
        sys.exit(1)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def _load_data(input_arg: str) -> dict:
    """Load overlay JSON from a file path or a raw JSON string."""
    if os.path.isfile(input_arg):
        with open(input_arg, 'r', encoding='utf-8') as fh:
            return json.load(fh)
    # Fallback: treat as raw JSON string
    return json.loads(input_arg)


if __name__ == '__main__':
    if len(sys.argv) < 4:
        print('Usage: python pdf_export.py <source.pdf> <output.pdf> '
              '<overlays.json | json-string>')
        sys.exit(1)

    source_pdf = sys.argv[1]
    output_pdf = sys.argv[2]
    data_arg   = sys.argv[3]

    try:
        overlay_data = _load_data(data_arg)
    except json.JSONDecodeError as exc:
        sys.stderr.write(f'[pdf_export] JSON parse error: {exc}\n')
        sys.exit(1)
    except Exception as exc:
        sys.stderr.write(f'[pdf_export] Could not load data: {exc}\n')
        sys.exit(1)

    process_pdf(source_pdf, output_pdf, overlay_data)

File: pdf-export.ts
import { Notice, Platform, TFile, FileSystemAdapter, normalizePath } from 'obsidian';
import OpenRouterTranslatorPlugin from './main';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** One translated region sent to the Python overlay script. */
interface ExportOverlay {
    text: string;
    page: number;
    x: number;
    y: number;
    width: number;
    height: number;
    fontFamily: string;
    /** Original font sizes observed in the source block (index 0 is used). */
    fontSize: number[];
    color: string;
    /** When true, x/y/width/height are in the range [0, 1] relative to page size. */
    isNormalized: boolean;
}

export interface ExportOptions {
    preserveOriginalText?: boolean;
    backgroundColor?: string;
    backgroundOpacity?: number;
    textColor?: string;
    autoOpen?: boolean;
    outputFileName?: string;
    pages?: number[];
}

// ---------------------------------------------------------------------------
// HTML → plain text (runs in the renderer process, no Node needed)
// ---------------------------------------------------------------------------

/**
 * Strip HTML tags from a string and convert block / line-break elements into
 * plain-text newlines.  Handles `<br>`, `<p>`, `<div>`, `<li>`, headings,
 * `&nbsp;`, `&amp;`, `&lt;`, `&gt;`, `&quot;`, and numeric character refs.
 *
 * Uses the browser's own DOMParser so no third-party library is required and
 * entity decoding is automatic.
 */
function htmlToPlainText(html: string): string {
    // Fast path: no tags present
    if (!html.includes('<')) return html;

    // Inject newlines before block-level elements so they survive innerText
    // normalisation.  We do this with simple string replacements to avoid
    // having to walk the DOM.
    const BLOCK_RE = /<\/?(p|div|li|tr|h[1-6]|br)\b[^>]*>/gi;
    const withNewlines = html.replace(BLOCK_RE, (match) => {
        // Self-closing br or opening block → newline before, closing block → newline after
        return '\n' + match;
    });

    // Use DOMParser for correct entity decoding and tag removal
    try {
        const parser = new DOMParser();
        const doc = parser.parseFromString(withNewlines, 'text/html');
        // innerText respects white-space better than textContent
        const text = (doc.body as HTMLElement).innerText ?? doc.body.textContent ?? '';
        // Collapse 3+ consecutive newlines → 2 and trim
        return text.replace(/\n{3,}/g, '\n\n').trim();
    } catch {
        // Fallback: naive tag-strip regex (should never be reached in modern browsers)
        return html
            .replace(/<br\s*\/?>/gi, '\n')
            .replace(/<\/?(p|div|li|h[1-6])[^>]*>/gi, '\n')
            .replace(/<[^>]+>/g, '')
            .replace(/&nbsp;/gi, ' ')
            .replace(/&amp;/gi, '&')
            .replace(/&lt;/gi, '<')
            .replace(/&gt;/gi, '>')
            .replace(/&quot;/gi, '"')
            .replace(/\n{3,}/g, '\n\n')
            .trim();
    }
}

// ---------------------------------------------------------------------------
// Service
// ---------------------------------------------------------------------------

export class PdfExportService {
    private plugin:     OpenRouterTranslatorPlugin;
    private scriptPath: string | null = null;
    private isDesktop:  boolean;

    // Lazily loaded Node.js / Electron modules
    private spawn: any = null;
    private path:  any = null;
    private fs:    any = null;
    private os:    any = null;

    constructor(plugin: OpenRouterTranslatorPlugin) {
        this.plugin    = plugin;
        this.isDesktop = Platform.isDesktop;

        if (this.isDesktop) {
            this.initDesktopModules();
        }
    }

    // -----------------------------------------------------------------------
    // Initialisation
    // -----------------------------------------------------------------------

    private initDesktopModules(): void {
        try {
            const nodeRequire = (window as any).require;
            if (!nodeRequire) {
                console.error('[PDF Export] window.require unavailable — Electron context missing?');
                this.isDesktop = false;
                return;
            }

            const cp   = nodeRequire('child_process');
            this.spawn = cp.spawn;
            this.path  = nodeRequire('path');
            this.fs    = nodeRequire('fs');
            this.os    = nodeRequire('os');

            const basePath  = this.getVaultBasePath();
            const pluginDir = this.plugin.manifest.dir ?? '';

            if (basePath) {
                this.scriptPath = this.path.join(basePath, pluginDir, 'pdf_export.py');
            } else {
                console.error('[PDF Export] Could not determine vault base path.');
                this.isDesktop = false;
            }
        } catch (err) {
            console.error('[PDF Export] Failed to load Node.js modules:', err);
            this.isDesktop = false;
        }
    }

    private getVaultBasePath(): string | null {
        const adapter = this.plugin.app.vault.adapter;
        return adapter instanceof FileSystemAdapter ? adapter.getBasePath() : null;
    }

    // -----------------------------------------------------------------------
    // Public entry point
    // -----------------------------------------------------------------------

    async exportFullPdf(): Promise<void> {
        if (!this.isDesktop) {
            new Notice('❌ PDF Export is only available on desktop.');
            return;
        }
        if (!this.spawn || !this.scriptPath) {
            new Notice('❌ PDF Export failed to initialise — check the console.');
            return;
        }

        try {
            // ── Validate active file ──────────────────────────────────────
            const file = this.plugin.app.workspace.getActiveFile();
            if (!file || file.extension !== 'pdf') {
                new Notice('❌ Please open a PDF file first.');
                return;
            }

            // ── Locate the linked translation Markdown file ───────────────
            const mdPath = this.plugin.pdfToMdMap.get(file.path);
            if (!mdPath) {
                new Notice('❌ No linked translation file found for this PDF.');
                return;
            }
            const mdFile = this.plugin.app.vault.getAbstractFileByPath(mdPath);
            if (!(mdFile instanceof TFile)) {
                new Notice(`❌ Translation file missing: ${mdPath}`);
                return;
            }

            // ── Parse overlays ────────────────────────────────────────────
            new Notice('⏳ Parsing translations…');
            const overlays      = await this.parseMarkdownStructure(mdFile);
            const totalOverlays = Object.values(overlays).reduce((n, a) => n + a.length, 0);

            if (totalOverlays === 0) {
                new Notice('❌ No translation blocks found in the linked file.');
                return;
            }

            // ── Build paths ───────────────────────────────────────────────
            const basePath = this.getVaultBasePath();
            if (!basePath) throw new Error('Vault base path not found.');

            const sourceAbsPath  = this.path.join(basePath, file.path);
            const outputFileName = `${file.basename}-translated.pdf`;
            const parentPath     = file.parent?.path ?? '';
            const outputRelPath  = normalizePath(`${parentPath}/${outputFileName}`);
            const outputAbsPath  = this.path.join(basePath, outputRelPath);

            // ── Run Python ────────────────────────────────────────────────
            new Notice(`🚀 Exporting ${totalOverlays} translation(s) to PDF…`);
            await this.executePythonExport(sourceAbsPath, outputAbsPath, overlays);

            // ── Wait for the file to appear in Obsidian's cache ───────────
            const appeared = await this.waitForFile(outputRelPath, 3000, 300);

            if (appeared) {
                new Notice(`✅ Export saved:\n${outputFileName}`, 6000);
                const exportedFile = this.plugin.app.vault.getAbstractFileByPath(outputRelPath);
                if (exportedFile instanceof TFile) {
                    this.plugin.app.workspace.getLeaf('tab').openFile(exportedFile);
                } else {
                    new Notice('File created — browse to it in the file explorer.');
                }
            } else {
                throw new Error('Python script exited cleanly but the output file was not found on disk.');
            }
        } catch (err: any) {
            console.error('[PDF Export]', err);
            new Notice(`❌ Export failed: ${err?.message ?? err}`);
        }
    }

    // -----------------------------------------------------------------------
    // Markdown parsing
    // -----------------------------------------------------------------------

    /**
     * T5.2: THE canonical storage parser is now the single reader of
     * `.translations.md`. The previous hand-rolled walker diverged from
     * storage.ts in three ways (multi-line translation joining, `<!-- empty -->`
     * handling, html stripping) — a file that rendered perfectly could
     * export differently from what the overlay showed.
     */
    private async parseMarkdownStructure(
        file: TFile,
    ): Promise<Record<string, ExportOverlay[]>> {
        const pdfFile = this.plugin.app.workspace.getActiveFile();
        const content = await this.plugin.app.vault.read(file);
        const saved = this.plugin.storage.parseMarkdownOverlay(
            content,
            (pdfFile && pdfFile.extension === 'pdf') ? pdfFile : file,
        );
        const exportData: Record<string, ExportOverlay[]> = {};
        if (!saved) return exportData;

        for (const [pageKey, items] of Object.entries(saved.pageOverlays)) {
            for (const item of items) {
                const text = htmlToPlainText(item.translatedText || '');
                if (!text) continue;
                const region = item.relativeRect;
                if (!exportData[pageKey]) exportData[pageKey] = [];
                exportData[pageKey].push({
                    text,
                    page: item.page,
                    x: region.left,
                    y: region.top,
                    width: region.width,
                    height: region.height,
                    fontFamily: this.normalizeFontFamily(item.fontFamily),
                    fontSize: Array.isArray(item.originalFontSizes) && item.originalFontSizes.length > 0
                        ? item.originalFontSizes
                        : [item.fontSize ?? 10],
                    color: '#000000',
                    isNormalized: true,
                });
            }
        }
        return exportData;
    }

    private normalizeFontFamily(ff?: string): string {
        if (!ff) return 'helvetica';
        const lower = ff.toLowerCase();
        if (lower.includes('mono') || lower.includes('courier')) return 'courier';
        if (lower.includes('serif') && !lower.includes('sans'))   return 'times';
        return 'helvetica';
    }

    // -----------------------------------------------------------------------
    // Python execution
    // -----------------------------------------------------------------------

    private async executePythonExport(
        sourceAbs: string,
        destAbs:   string,
        data:      Record<string, ExportOverlay[]>,
    ): Promise<void> {
        if (!this.spawn || !this.path || !this.fs || !this.os) {
            throw new Error('Node.js modules not initialised.');
        }

        return new Promise<void>((resolve, reject) => {
            const pythonCmd = this.plugin.settings.pythonPath
                || (Platform.isWin ? 'python' : 'python3');

            // Write overlay data to a temp file with explicit UTF-8 encoding
            const tempFile = this.path.join(
                this.os.tmpdir(),
                `obsidian-pdf-export-${Date.now()}.json`,
            );

            try {
                this.fs.writeFileSync(tempFile, JSON.stringify(data), { encoding: 'utf-8' });
            } catch (err: any) {
                return reject(new Error(`Failed to write temp JSON: ${err.message}`));
            }

            console.log(`[PDF Export] Spawning: ${pythonCmd} "${this.scriptPath}"`);

            // Phase 4 (C1): `process.env` is a Node-only global — on non-
            // desktop / sandboxed renderer contexts it may be undefined.
            // We are already inside `exportFullPdf()` which is gated by
            // `this.isDesktop` (Platform.isDesktop), so spawning itself is
            // safe; we just need to access env safely. Use the Electron
            // `window.process` proxy when available, otherwise spawn with
            // an empty env (Python will still inherit a minimal env from
            // the Electron main process in practice).
            const spawnEnv = Platform.isDesktop
                ? { env: { ...(((window as any).process?.env) || {}) } }
                : {};

            const child = this.spawn(
                pythonCmd,
                [this.scriptPath, sourceAbs, destAbs, tempFile],
                spawnEnv,   // copy env so PATH is inherited (desktop only)
            );

            let stdout = '';
            let stderr = '';

            child.stdout.on('data', (chunk: any) => { stdout += chunk.toString(); });
            child.stderr.on('data', (chunk: any) => { stderr += chunk.toString(); });

            child.on('close', (code: number) => {
                this.tryDeleteTemp(tempFile);

                if (stdout) console.log('[PDF Export] stdout:\n' + stdout);
                if (stderr) console.warn('[PDF Export] stderr:\n' + stderr);

                if (code === 0) {
                    resolve();
                } else {
                    const msg = this.extractPythonError(stderr) || `Exit code ${code}`;
                    reject(new Error(msg));
                }
            });

            child.on('error', (err: any) => {
                this.tryDeleteTemp(tempFile);
                reject(new Error(
                    `Could not start Python ("${pythonCmd}"): ${err.message}. ` +
                    'Check the Python Path setting.',
                ));
            });
        });
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    /**
     * Poll for a vault file to appear, up to ``timeoutMs`` milliseconds,
     * checking every ``intervalMs`` milliseconds.
     */
    private async waitForFile(
        relPath:    string,
        timeoutMs:  number,
        intervalMs: number,
    ): Promise<boolean> {
        const deadline = Date.now() + timeoutMs;
        while (Date.now() < deadline) {
            await new Promise(r => setTimeout(r, intervalMs));
            if (await this.plugin.app.vault.adapter.exists(relPath)) {
                return true;
            }
        }
        return false;
    }

    /** Extract the most informative line from a Python traceback. */
    private extractPythonError(log: string): string {
        const lines = log.split('\n').map(l => l.trim()).filter(Boolean);
        // Walk backwards; prefer lines with "Error:" or "Exception:"
        for (let i = lines.length - 1; i >= 0; i--) {
            if (lines[i].match(/Error:|Exception:/)) return lines[i];
        }
        // Fallback: last non-empty line
        return lines[lines.length - 1] ?? 'Unknown Python error — check the console.';
    }

    private tryDeleteTemp(filePath: string): void {
        try {
            if (this.fs?.existsSync(filePath)) this.fs.unlinkSync(filePath);
        } catch { /* best-effort */ }
    }
}
